File: src/lira/vocabulary/data/lexical_relationship_tensor.py
"""LexicalRelationship system-properties tensor store: one growable row
per LexicalRelationship, holding its confidence/provenance/temporal/
activation weights in one dense array, grown by doubling (Rule 14) --
the same discipline as LinguisticSystemPropertyTensor in the
Linguistics Layer. Only LexicalRelationship gets a row here; Dictionary
and Word do not (Design Principle 8: tensor-backed system properties
apply only to word relationships, not to words standing alone).

uuid and version are non-numeric per-row data and live in plain Python
lists alongside the tensor, mirroring LinguisticSystemPropertyTensor's
_uuids/_origins convention -- never packed into the tensor itself."""

import numpy as np

CONFIDENCE_COL = 0
PROVENANCE_COL = 1
TEMPORAL_COL = 2
ACTIVATION_COL = 3
N_COLS = 4


class LexicalRelationshipSystemPropertyTensor:
    def __init__(self, initial_capacity: int = 16):
        self._capacity = initial_capacity
        self._n_rows = 0
        self.values = np.zeros((self._capacity, N_COLS), dtype=np.float64)
        self._uuids = []
        self._versions = []

    def _grow(self):
        new_capacity = self._capacity * 2
        new_values = np.zeros((new_capacity, N_COLS), dtype=self.values.dtype)
        new_values[:self._n_rows] = self.values[:self._n_rows]
        self.values = new_values
        self._capacity = new_capacity

    def allocate_row(self, uuid_str: str, version: str, confidence: float = 0.0,
                      provenance: float = 0.0, temporal: float = 0.0, activation: float = 0.0) -> int:
        if self._n_rows >= self._capacity:
            self._grow()
        row = self._n_rows
        self._n_rows += 1

        self.values[row, CONFIDENCE_COL] = confidence
        self.values[row, PROVENANCE_COL] = provenance
        self.values[row, TEMPORAL_COL] = temporal
        self.values[row, ACTIVATION_COL] = activation

        self._uuids.append(uuid_str)
        self._versions.append(version)
        return row

    def uuid_of(self, row: int) -> str:
        return self._uuids[row]

    def version_of(self, row: int) -> str:
        return self._versions[row]

File: prototype/src/lira/vocabulary/data/lexical_relationship_tensor.ts
import { ACTIVATION_COL, CONFIDENCE_COL, N_COLS, PROVENANCE_COL, TEMPORAL_COL } from "./morphological_pointer_relationship_tensor";

/** LexicalRelationship system-properties tensor store --
 * `SemanticRelationshipSystemPropertyTensor`'s own exact counterpart
 * (data/semantic_relationship_tensor.ts), one growable row per
 * (permanent) `LexicalRelationship` instead. Shares
 * `MorphologicalPointerRelationshipSystemPropertyTensor`'s own column
 * layout, the same reason `SemanticRelationshipSystemPropertyTensor`
 * does (Design Principle 8 -- tensor-backed system properties apply
 * only to relationship records, not to words/senses/forms standing
 * alone). */
export class LexicalRelationshipSystemPropertyTensor {
  private capacity: number;
  private nRows = 0;
  values: Float64Array;
  private uuids: string[] = [];
  private versions: string[] = [];

  constructor(initialCapacity = 16) {
    this.capacity = initialCapacity;
    this.values = new Float64Array(this.capacity * N_COLS);
  }

  private grow(): void {
    const newCapacity = this.capacity * 2;
    const newValues = new Float64Array(newCapacity * N_COLS);
    newValues.set(this.values.subarray(0, this.nRows * N_COLS));
    this.values = newValues;
    this.capacity = newCapacity;
  }

  private cell(row: number, col: number): number {
    return row * N_COLS + col;
  }

  allocateRow(
    uuidStr: string,
    version: string,
    confidence = 0.0,
    provenance = 0.0,
    temporal = 0.0,
    activation = 0.0,
  ): number {
    if (this.nRows >= this.capacity) this.grow();
    const row = this.nRows;
    this.nRows += 1;

    this.values[this.cell(row, CONFIDENCE_COL)] = confidence;
    this.values[this.cell(row, PROVENANCE_COL)] = provenance;
    this.values[this.cell(row, TEMPORAL_COL)] = temporal;
    this.values[this.cell(row, ACTIVATION_COL)] = activation;

    this.uuids.push(uuidStr);
    this.versions.push(version);
    return row;
  }

  getCell(row: number, col: number): number {
    return this.values[this.cell(row, col)];
  }

  setCell(row: number, col: number, value: number): void {
    this.values[this.cell(row, col)] = value;
  }

  uuidOf(row: number): string {
    return this.uuids[row];
  }

  versionOf(row: number): string {
    return this.versions[row];
  }
}
